src/main/java/seedu/address/model/VersionedAddressBook.java

```diff
@@ -0,0 +1,156 @@
+    }
+
+    /**
+     * Reverts the given AddressBook to the previous state.
+     */
+    public void undoAddressBook() throws CommandException {
+        if (currentStatePointer <= 0) {
+            throw new CommandException(MESSAGE_NO_MORE_HISTORY);
+        }
+
+        if (!current.equals(addressBookStateList.get(currentStatePointer))) {
+            throw new CommandException(MESSAGE_UNSAVED_CHANGES);
+        }
+
+        currentStatePointer--;
+        current = new AddressBook(addressBookStateList.get(currentStatePointer));
+    }
+
+    /**
+     * Discards the unsaved changes in the current state.
+     */
+    public void discardUnsavedChanges() {
+        current = new AddressBook(addressBookStateList.get(currentStatePointer));
+    }
+
+    /**
+     * Return an unmodifiable list containing copies of the AddressBook instances for testing purpose.
+     */
+    public ArrayList<AddressBook> getAddressBookStateList() {
+        ArrayList<AddressBook> copy = new ArrayList<>();
+        for (AddressBook addressBook : addressBookStateList) {
+            copy.add(new AddressBook(addressBook)); // Assuming AddressBook is immutable
+        }
+        return copy;
+    }
+
+    @Override
+    public void setPersons(List<Person> persons) {
+        getCurrentAddressBook().setPersons(persons);
+    }
+
+    @Override
+    public void resetData(ReadOnlyAddressBook newData) {
+        getCurrentAddressBook().resetData(newData);
+    }
+
+    @Override
+    public boolean hasPerson(Person person) {
+        return getCurrentAddressBook().hasPerson(person);
+    }
+
+    @Override
+    public void addPerson(Person p) {
+        getCurrentAddressBook().addPerson(p);
+    }
+
+    @Override
+    public void setPerson(Person target, Person editedPerson) {
+        getCurrentAddressBook().setPerson(target, editedPerson);
+    }
+
+    @Override
+    public void removePerson(Person key) {
+        getCurrentAddressBook().removePerson(key);
+    }
+
+    @Override
+    public ObservableList<Person> getPersonList() {
+        return getCurrentAddressBook().getPersonList();
+    }
+
+    @Override
+    public String toString() {
+        return getCurrentAddressBook().toString();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null) {
+            return false;
+        }
+        if (!(o instanceof VersionedAddressBook)) {
+            return false;
+        }
+        VersionedAddressBook other = (VersionedAddressBook) o;
+
+        // Compare addressBookStateList and currentStatePointer
+        return currentStatePointer == other.currentStatePointer
+                && addressBookStateList.equals(other.addressBookStateList)
+                && current.equals(other.current);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(addressBookStateList, currentStatePointer);
```

Comment:
Since we are comparing the ``current`` instance in the equals method, would it be better if we consider ``current`` in the hash code as well?

Reply:
Good catch! Although the current code would work fine (it satisfies the implication required for hashing: `vab1.equals(vab2) => vab1.hashCode() == vab2.hashCode()`) but would cause quite a lot of hash collisions. Adding `current` in `hashCode()` is indeed better!